pkg/operators/ebpf/stats.go

```diff
@@ -0,0 +1,535 @@
+// Copyright 2025 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpfoperator
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/spf13/viper"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/bpfstats"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	metadatav1 "github.com/inspektor-gadget/inspektor-gadget/pkg/metadata/v1"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+const (
+	ParamGadgetStatisticsInterval = "statistics-interval"
+	ParamGadgetStatisticsCount    = "statistics-count"
+	EmitStatsAnn                  = "emitstats"
+	ParamProgramStats             = "programs"
+	StatsDSName                   = "bpfstats"
+	EnableBPFStatsParam           = "enable-bpfstats"
+)
+
+func (i *ebpfOperator) GlobalParams() api.Params {
+	return api.Params{
+		{
+			Key:          EnableBPFStatsParam,
+			Description:  "Enable capturing eBPF stats",
+			DefaultValue: "false",
+			TypeHint:     api.TypeBool,
+			Title:        "Enable eBPF stats",
+		},
+	}
+}
+
+func (i *ebpfOperator) Init(params *params.Params) error {
+	if !params.Get(EnableBPFStatsParam).AsBool() {
+		return nil
+	}
+
+	i.gadgetObjs = make(map[operators.GadgetContext]gadgetObjs)
+
+	// Enable stats collection
+	// TODO: when to disable it?
```

Comment:
I think it should be enabled when starting the data operator instance and disabled when stopping it. Then, the internal `refCnt` of bpfstats package should do the trick.

BTW, I think the logic is missing an increasement when it has been already enabled:

```go
func EnableBPFStats() error {
	mutex.Lock()
	defer mutex.Unlock()

	if refCnt != 0 {
		refCnt++ // This is missing in the original code
		return nil
	}
```

Or, did I miss something?

Reply:
> I think it should be enabled when starting the data operator instance and disabled when stopping it.

If the option is enabled globally (--enable-bpfstats), we need to enable it when the daemon starts, otherwise we'll miss the stats for gadgets run before the stats command/gadget.

The issue here is that operators don't have a Close() or similar method to perform cleanup tasks.

Reply:
> BTW, I think the logic is missing an increasement when it has been already enabled:
> 
> ```go
> func EnableBPFStats() error {
> 	mutex.Lock()
> 	defer mutex.Unlock()
> 
> 	if refCnt != 0 {
> 		refCnt++ // This is missing in the original code
> 		return nil
> 	}
> ```
> 
> Or, did I miss something?

Yes, that seems like a bug. It's fixed in #3976

Reply:
> If the option is enabled globally (--enable-bpfstats), we need to enable it when the daemon starts, otherwise we'll miss the stats for gadgets run before the stats command/gadget.
>
Aaahh because we want to collect stats from the moment a gadget starts, not from the moment the stats command/gadget is executed. Good point.

> The issue here is that operators don't have a Close() or similar method to perform cleanup tasks.
>
Yeah, then I agree.